packages/utils/src/schema/sanitizeDataForNewSchema.ts

```diff
@@ -0,0 +1,208 @@
+      const properties = get(oldSchema, PROPERTIES_KEY, {});
+      Object.keys(properties).forEach((key) => {
+        if (has(data, key)) {
+          removeOldSchemaData[key] = undefined;
+        }
+      });
+    }
+    const keys: string[] = Object.keys(get(newSchema, PROPERTIES_KEY, {}));
+    // Create a place to store nested data that will be a side-effect of the filter
+    const nestedData: GenericObjectType = {};
+    keys.forEach((key) => {
+      const formValue = get(data, key);
+      let oldKeyedSchema: S = get(oldSchema, [PROPERTIES_KEY, key], {});
+      let newKeyedSchema: S = get(newSchema, [PROPERTIES_KEY, key], {});
+      // Resolve the refs if they exist
+      if (has(oldKeyedSchema, REF_KEY)) {
+        oldKeyedSchema = retrieveSchema<T, S, F>(
+          validator,
+          oldKeyedSchema,
+          rootSchema,
+          formValue
+        );
+      }
+      if (has(newKeyedSchema, REF_KEY)) {
+        newKeyedSchema = retrieveSchema<T, S, F>(
+          validator,
+          newKeyedSchema,
+          rootSchema,
+          formValue
+        );
+      }
+      // Now get types and see if they are the same
+      const oldSchemaTypeForKey = get(oldKeyedSchema, "type");
+      const newSchemaTypeForKey = get(newKeyedSchema, "type");
+      // Check if the old option has the same key with the same type
+      if (!oldSchemaTypeForKey || oldSchemaTypeForKey === newSchemaTypeForKey) {
+        if (has(removeOldSchemaData, key)) {
+          // SIDE-EFFECT: remove the undefined value for a key that has the same type between the old and new schemas
+          delete removeOldSchemaData[key];
+        }
+        // If it is an object, we'll recurse and store the resulting sanitized data for the key
+        if (
+          newSchemaTypeForKey === "object" ||
+          (newSchemaTypeForKey === "array" && Array.isArray(formValue))
+        ) {
+          // SIDE-EFFECT: process the new schema type of object recursively to save iterations
+          const itemData = sanitizeDataForNewSchema<T, S, F>(
+            validator,
+            rootSchema,
+            newKeyedSchema,
+            oldKeyedSchema,
+            formValue
+          );
+          if (itemData !== undefined || newSchemaTypeForKey === "array") {
+            // only put undefined values for the array type and not the object type
+            nestedData[key] = itemData;
+          }
+        } else {
+          // Ok, the non-object types match, let's make sure that a default or a const of a different value is replaced
+          // with the new default or const
+          const newOptionDefault = get(newKeyedSchema, ["default"]);
+          const newOptionConst = get(newKeyedSchema, ["const"]);
+          if (
+            (newOptionDefault && newOptionDefault !== formValue) ||
+            (newOptionConst && newOptionConst !== formValue)
+          ) {
+            removeOldSchemaData[key] = newOptionDefault;
```

Comment:
Making this undefined actually didn't fix one or two of the issues. In this case two schemas had a single named property but each had a different default value. When you changed from the first to the second, clearing the value here kept the first selected because getDefaultFormState's use of `getFirstMatchingOption` would choose the first option again. By switching from one default to another, it picked the second.